### DISCUSSION TITLE
Load static HTML files

### DISCUSSION THREAD

**Discussion by robinzimmer1989** [vercel/next.js#21666] [Help] [answered]
I'm working on a project at the moment where one requirement is to dynamically load static HTML pages. Those pages must be protected, so I can't just place them in the public folder.

My idea was to move them to the root and then use getServerSideProps to read the slug and search for the file via readFileSync. It's working in development and it's partially working in production.

This is my folder structure:
```
docs
-index.html
-some-folder
--some-file.html
pages
-docs
--[...slug].js
--index.js
```

This is index.js in pages/docs:
```
import React from "react"
import { readFileSync } from "fs"

import Document from "../../components/Document"

const Docs = (props) => {
  return <Document html={props.html}  />
}

export async function getServerSideProps() {
  const html = readFileSync(`./docs/index.html`, "utf8")

  return {
    props: { html },
  }
}

export default Docs
```

And this is [...slug].js in pages/docs:
```
import React from "react"
import { readFileSync } from "fs"
import Document from "../../components/Document"

const Doc = (props) => {
  return <Document html={props.html} />
}

export async function getServerSideProps({ params: { slug } }) {
  const formattedSlug = Array.isArray(slug) ? slug.join("/") : slug
  const html = readFileSync(`./docs/${formattedSlug}.html`, "utf8")

  return {
    props: { html },
  }
}

export default Doc
```

It's working in development and it's working in production for index.html but not for the sub-pages generated by the [...slug].js file. Here I get the following error message:
```
An error occurred with this application.
This is an error with the application itself, not the platform.

502: BAD_GATEWAY
Code: NO_RESPONSE_FROM_FUNCTION

```
And this is what the function log spits out:
```
ERROR	Unhandled error during request: Error: ENOENT: no such file or directory, open '/vercel/workpath0/docs/test.html'
```

But the test.html file exists in the /docs folder.

Did I miss anything? I'm pretty new to NextJS so maybe there is an even better way to achieve what I'm trying to do?

Thanks for your help!

Robin

**Comment by michrome** [comment, marked as answer]
Does it work (for test.html) if you hard code it? E.g. replace:

```js
const html = readFileSync(`./docs/${formattedSlug}.html`, "utf8")
```

With:

```js
const html = readFileSync(`./docs/test.html`, "utf8")
```

?

I’m wondering if there’s something funky going on with the dynamic file path.

**Reply by robinzimmer1989** [reply]
Good catch, hardcoding the file works!

How do I debug now? I've tried to use console.log and hoped to see the result in the functions log but without luck. Is there a server-side log function available?

Thanks!

**Reply by michrome** [reply]
This relates to https://github.com/vercel/next.js/issues/8251 . The workaround is to use `require`.

If you have this directory structure:

```
docs
-test.txt
-test2.txt
pages
-loader
--[slug].js
```

Then you can do something like:

```js
// pages/loader/[slug].js
import { readFileSync } from "fs"
const { join } = require("path")

export default function Page({ content }) {
  return <>{content}</>
}

export async function getServerSideProps({ params: { slug } }) {
  const file = join(process.cwd(), "docs", slug)
  const content = readFileSync(file, "utf8")
  return {
    props: { content },
  }
}
```

e.g.
* https://j4e-name-quwhw5we3.vercel.app/loader/test.txt
* https://j4e-name-quwhw5we3.vercel.app/loader/test2.txt

**Reply by robinzimmer1989** [reply]
Perfect, 'require' fixed it!

Thanks for your help!
